src/bms.lisp

```diff
@@ -0,0 +1,49 @@
+(defun bms-loop () {
+    (var bms-vmin-limit-start (conf-get 'bms-vmin-limit-start))
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; Start 3 degrees before Motor CFG -> BMS limiting functionality would happen.
+    (var bms-tmin-limit-start 0)
+    (var bms-cell-balance-start 0.5)
+    (var bms-ic-tmax-limit-start (+ bms-tmax-limit-start 15)) ; Set bms temp limit to +15C past cell limit
+    (var bms-config-update-time (systime))
+    (var bms-fault 0u32)
+    (var bms-fault-codes '(
+        (NONE . 0)
+        (BMS_CONNECTION . 1)
+        (BMS_OVER_TEMP . 2)
+        (BMS_CELL_OVER_VOLTAGE . 3)
+        (BMS_CELL_UNDER_VOLTAGE . 4)
+        (BMS_CELL_OVER_TEMP . 5)
+        (BMS_CELL_UNDER_TEMP . 6)
+        (BMS_CELL_BALANCE . 7)
+    ))
+    (defun bms-set-fault (fault-code) {
+        (if (eq fault-code 'NONE) {
+            (setq bms-fault 0u32)
+        }{
+            (setq bms-fault (bitwise-or bms-fault (shl 1 (- (assoc bms-fault-codes fault-code) 1))))
+        })
+    })
+    (loopwhile t {
+        (if (and (<= (get-duty) 0.05)  (> (secs-since bms-config-update-time) 1.0)) { ; Only bother updating config values while board is idle
+            (setq bms-vmin-limit-start (conf-get 'bms-vmin-limit-start))
+            (setq bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+            (setq bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3))
+            (setq bms-ic-tmax-limit-start (+ bms-tmax-limit-start 15))
+            (setq bms-config-update-time (systime))
+        })
```

Comment:
Unaddressed from previous review, this code is duplicate with the one on lines 2-7. Please move it out to a function.

Reply:
Technically this is not duplicate. Lines 2-7 are variable declaration (var) and initial assignment. I'll move it to a separately function, but technically that's just creating more overhead. Lisp is not a compiled language...

Reply:
An overhead of a single function call? Unless that's more expensive than I though, I just consider it a necessity for well-written code, it's significantly better from maintenance perspective to not duplicate e.g. the 3 and the 15 constants.